### ISSUE TITLE
Update Helm chart with current best practices - PRs wanted?

### ISSUE THREAD

**Issue by consideRatio** [kubernetes-sigs/nfs-ganesha-server-and-external-provisioner#19]
As Helm has developed over time, various best practices has been updated. I'd be happy to contribute PRs that would make this Helm chart up to date with such practices. But, I'd like some input with regards to if that would be wanted, and if it could be okay to make some breaking changes at this time.

I figure since this Helm chart isn't being published right now, it would be suitable to do breaking changes now before it becomes published again, for example at https://artifacthub.io/ which is the new Helm Hub location.

## Breaking changes examplified
- _Use modern labels, so instead of `app` it would be `app.kubernetes.io/name`  (like discussed for this Helm chart: https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/1867)._
  This is tricky to update with `helm3 upgrade` even though it was quite easy with `helm2 upgrade --force`, so if it is ever to be done, it's good to do sooner rather than later in my mind.
- _Apply 
- _Update Chart.yaml to `apiVersion: 2`._
  A change that will make Helm 2 no longer be supported, but Helm 2 reach end of life November 13.

## Non breaking changes examplified
- Apply systematic whitespace chomping practices (always chomp left, almost never chomp right).
- Use `{{- with ... }}` instead of `{{- if }}` when it makes sense
- Use `include` instead of `template`.
- Use `nindent` with whitespace chomping left instead of `indent` when using `toYaml`
- Follow indentation practices for readability

## Feedback wanted
Would you be open to such breaking changes and/or such non breaking changes to the Helm chart?

## About me
I'm maintaining the [JupyterHub Helm chart](https://github.com/jupyterhub/zero-to-jupyterhub-k8s) and have spent a lot of time working with k8s and Helm chart the last ~3 years. I'm motivated to contribute to this project as I think it serves a very important piece in the infrastructure I want to help many others make use of, for example universities that want to provide a compute platform for their researchers and students with personal home directories.

**Comment by kmova** [comment]
cc: @kvaps 

@consideRatio thank you for posting this and also offering to help. It would be great to start with the non-breaking changes and then look at possible breaking changes PR as well.

**Comment by kvaps** [comment]
Hi @consideRatio, we're always happy about any contribution, I totally agree with the following changes:


> 
>   * Apply systematic whitespace chomping practices (always chomp left, almost never chomp right).
> 
>   * Use `{{- with ... }}` instead of `{{- if }}` when it makes sense
> 
>   * Use `nindent` with whitespace chomping left instead of `indent` when using `toYaml`
> 
>   * Follow indentation practices for readability


Please correct me if I', wrong, but `include` is usually makes sense in case when you need to save the templated output into variable, it useless for the single injection.

>   * Use `include` instead of `template`.

I don't see big reason to apply the following changes, despite the fact the Helm does not providing seamless upgrade:

>   * _Use modern labels, so instead of `app` it would be `app.kubernetes.io/name`  (like discussed for this Helm chart: [jupyterhub/zero-to-jupyterhub-k8s#1867](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/1867))._
>       This is tricky to update with `helm3 upgrade` even though it was quite easy with `helm2 upgrade --force`, so if it is ever to be done, it's good to do sooner rather than later in my mind.
> 
>   * _Apply
> 
>   * _Update Chart.yaml to `apiVersion: 2`._
>       A change that will make Helm 2 no longer be supported, but Helm 2 reach end of life November 13.

What did you mean by `_Apply`, typo?

I would also add an opportunity to specify the digest for the docker image, see the similar commit for nginx-ingress chart
https://github.com/kubernetes/ingress-nginx/commit/2205edb16b0a27b7eb84d86422d3fb9cc8bef715

However I believe that there is better syntax:

```mustage
{{- with .Values.image }}
image: "{{ .repository }}{{ if .digest }}@{{ .digest }}{{ else }}:{{ .tag }}{{ end }}"
imagePullPolicy: {{ .PullPolicy }}
{{- end }}
```

> I figure since this Helm chart isn't being published right now, it would be suitable to do breaking changes now before it becomes published again, for example at https://artifacthub.io/ which is the new Helm Hub location.

It is published in my helm charts repository
https://artifacthub.io/packages/helm/kvaps/nfs-server-provisioner

**Comment by consideRatio** [comment]
Thanks for the followup @kmova and @kvaps! I opened #21 with non breaking refactoring only.

---

> Please correct me if I', wrong, but include is usually makes sense in case when you need to save the templated output into variable, it useless for the single injection.

@kvaps I didn't understand "the single injection." reference, but I think we are in agreement. As I understand it, `include` can do what `template` can and more, such as allowing its output to set a variable or be piped into another function.

The templates generated by `helm create` currently only use `include` instead of `template`. I recall the reason for that was to consistently use `include` over `template` to avoid having anyone need to learn the difference as there was no downside to always using include.

I added a commit with this change, but of course git history will be clutter a bit, so let me know if you want me to remove it again and I'll quickly rebase it to oblivion.

Here is a note about include vs template [from the Helm docs](https://helm.sh/docs/chart_template_guide/named_templates/#the-include-function):

![image](https://user-images.githubusercontent.com/3837114/100387226-d9bf9000-3027-11eb-8e91-ef4521d7ed36.png)


> What did you mean by _Apply, typo?

Yepp this was a typo!
